### ISSUE TITLE
endpoints fail with terraform 0.12.29 (which is allowed) - cannot run plan&apply of complete example

### ISSUE THREAD

**Issue by tbugfinder** [terraform-aws-modules/terraform-aws-vpc#646]
https://github.com/terraform-aws-modules/terraform-aws-vpc/blob/73b447ff1d6d6b7bda2adf45ccfba6ae9cd1d832/modules/vpc-endpoints/main.tf#L2

```
$ terraform -version
Terraform v0.12.29

###

terraform init
terraform plan
Error: Inconsistent conditional result types

  on ../../modules/vpc-endpoints/main.tf line 2, in locals:
   2:   endpoints = var.create ? var.endpoints : tomap({})
    |----------------
    | var.create is true
    | var.endpoints is object with 14 attributes

The true and false result expressions must have consistent types. The given
expressions are object and map of dynamic, respectively.


```

**Comment by marcohenriques** [comment]
I'm also having this issue in terraform version 0.15.3 when using vpc-endpoints submodule

**Comment by Erokos** [comment]
Hi, sorry to intrude but can you please help me with specifying the source parameter for the vpc-endpoints submodule? I just can't get it right...

**Comment by rmchale** [comment]
I am also having this issue.

**Comment by tbugfinder** [comment]
I did some research on different options to fix this issue, however I couldn't find anything which just fixes above report line of code, so it would require more changes.

I think maitainers should add some thought here to. Do you want to drop terraform 0.12.x support or should for_each be rewritten for vpc_endpoints?

**Comment by dctrwatson** [comment]
Workaround adapted from complete example:
```
locals {
  endpoints = {
    s3 = {
      service = "s3"
      tags    = { Name = "s3-vpc-endpoint" }
    },
    dynamodb = {
      service         = "dynamodb"
      service_type    = "Gateway"
      route_table_ids = flatten([module.vpc.intra_route_table_ids, module.vpc.private_route_table_ids, module.vpc.public_route_table_ids])
      policy          = data.aws_iam_policy_document.dynamodb_endpoint_policy.json
      tags            = { Name = "dynamodb-vpc-endpoint" }
    },
    ssm = {
      service             = "ssm"
      private_dns_enabled = true
      subnet_ids          = module.vpc.private_subnets
    },
    ssmmessages = {
      service             = "ssmmessages"
      private_dns_enabled = true
      subnet_ids          = module.vpc.private_subnets
    },
  ]

  create_endpoints = false
}

module "vpc_endpoints" {
  source = "../../modules/vpc-endpoints"

  create = local.create_endpoints

  vpc_id             = module.vpc.vpc_id
  security_group_ids = [data.aws_security_group.default.id]
  
  endpoints = { for service, ep in local.endpoints : (service) => ep if local.create_endpoints }
  
  tags = merge(local.tags, {
    Project  = "Secret"
    Endpoint = "true"
  })
}
```

**Comment by ejimz** [comment]
Same here with v0.14.11

**Comment by SkypLabs** [comment]
Same issue with Terraform v1.0.1.

**Comment by bryantbiggs** [comment]
fixed in #729
